### ISSUE TITLE
only support BLE over SPI/SDIO?

### ISSUE THREAD

**Issue by felix-lhp** [espressif/esp-hosted#295]
@esp_hosted_ng/host/main.c:
```
void process_capabilities(struct esp_adapter *adapter)
{
	esp_info("ESP peripheral capabilities: 0x%x\n", adapter->capabilities);

	/* Reset BT */
	esp_deinit_bt(adapter);

	if ((adapter->capabilities & ESP_BT_SPI_SUPPORT) ||
		(adapter->capabilities & ESP_BT_SDIO_SUPPORT)) {
		msleep(200);
		esp_info("ESP Bluetooth init\n");
		esp_init_bt(adapter);
	}
}
```

the startup log @esp32c3
```
I (378) coexist: coexist rom version 9387209
I (383) cpu_start: Starting scheduler.
I (388) stats: *********************************************************************
I (388) stats:                 ESP-Hosted Firmware version :: 1.0.2                        
I (398) stats:                 Transport used :: SPI + UART                    
I (408) stats: *********************************************************************
I (418) FW_MAIN: Supported features are:
I (418) FW_MAIN: - WLAN over SPI
I (428) FW_BT: - BT/BLE
I (428) FW_BT:    - HCI Over UART
I (428) FW_BT:    - BLE only
I (438) FW_MAIN: Capabilities: 0xaa
I (448) pp: pp rom version: 9387209
I (448) net80211: net80211 rom version: 9387209
I (458) wifi:wifi driver task: 3fc9e6ec, prio:23, stack:6656, core=0
I (458) system_api: Base MAC address is not set
I (458) system_api: read default base MAC address from EFUSE
I (468) wifi:wifi firmware version: 9f5e450
I (468) wifi:wifi certification version: v7.0
```

the dmesg @linux:
```
[    3.534666] venc_release 50
[    4.486794] esp32c3:process_esp_bootup_event: 
[    4.486794] Received ESP bootup event
[    4.486855] esp32c3:process_event_esp_bootup: EVENT: 3
[    4.486876] esp32c3:process_event_esp_bootup: EVENT: 2
[    4.486897] esp32c3:adjust_spi_clock: ESP Reconfigure SPI CLK to 30 MHz
[    4.486915] esp32c3:process_event_esp_bootup: EVENT: 0
[    4.486933] esp32c3:process_event_esp_bootup: EVENT: 4
[    4.486951] esp32c3:process_event_esp_bootup: EVENT: 1
[    4.486971] esp32c3:process_fw_data: ESP chipset's last reset cause:
[    4.486988] esp32c3:print_reset_reason: POWERON_RESET
[    4.487009] esp32c3:check_esp_version: esp32: ESP Firmware version: 1.0.2
[    4.487031] esp32c3:process_event_esp_bootup: ESP chipset detected [esp32-c3]
[    4.487103] esp32c3:add_network_iface: add_network_iface
[    4.487599] esp32c3:esp_reg_notifier: 629 esp_wifi_device not initialized yet 
[    4.487871] esp32c3:add_network_iface: add_network_iface before esp_cfg80211_add_iface
[    4.527298] esp32c3:add_network_iface: add_network_iface after esp_cfg80211_add_iface
[    4.527342] esp32c3:process_capabilities: ESP peripheral capabilities: 0xaa
[    4.527368] esp32c3:print_capabilities: Capabilities: 0xaa. Features supported are:
[    4.527386] esp32c3:print_capabilities:       * WLAN on SPI
[    4.527401] esp32c3:print_capabilities:       * BT/BLE
[    4.527418] esp32c3:print_capabilities:         - HCI over UART
[    4.527435] esp32c3:print_capabilities:         - BLE only
[    8.274566] esp32c3:cmd_auth_request: Authentication request: f4:84:8d:fb:5a:17 6 0 0 0
[    8.517273] esp32c3:cmd_assoc_request: Association request: f4:84:8d:fb:5a:17 6 22
[    8.548892] esp32c3:process_assoc_event: Connection status: 0
[    8.549930] esp32c3:process_rx_packet: Rx PACKET_TYPE_EAPOL!!!!
[    8.550514] NOHZ tick-stop error: Non-RCU local softirq work is pending, handler #08!!!
[    8.566302] esp32c3:process_rx_packet: Rx PACKET_TYPE_EAPOL!!!!
[    8.566430] NOHZ tick-stop error: Non-RCU local softirq work is pending, handler #08!!!
[   10.706334] esp32c3:esp_inetaddr_event: NETDEV_UP interface espsta0 ip changed to  192.168.001.138
```

So,how can I support ble over uart?

**Comment by felix-lhp** [comment]
Sorry, Let me edit the question again:

`@esp_hosted_ng/host/main.c:
void process_capabilities(struct esp_adapter *adapter)
{
esp_info("ESP peripheral capabilities: 0x%x\n", adapter->capabilities);

/* Reset BT */
esp_deinit_bt(adapter);

if ((adapter->capabilities & ESP_BT_SPI_SUPPORT) ||
	(adapter->capabilities & ESP_BT_SDIO_SUPPORT)) {
	msleep(200);
	esp_info("ESP Bluetooth init\n");
	esp_init_bt(adapter);
}
}`

the start log@esp32c3:
I (365) sleep: Configure to isolate all GPIO pins in sleep state
I (371) sleep: Enable automatic switching of GPIO sleep configuration
I (378) coexist: coexist rom version 9387209
I (383) cpu_start: Starting scheduler.
I (388) stats: *********************************************************************
I (388) stats:                 ESP-Hosted Firmware version :: 1.0.2                        
I (398) stats:                 Transport used :: SPI + UART                    
I (408) stats: *********************************************************************
I (418) FW_MAIN: Supported features are:
I (418) FW_MAIN: - WLAN over SPI
I (428) FW_BT: - BT/BLE
I (428) FW_BT:    - HCI Over UART
I (428) FW_BT:    - BLE only
I (438) FW_MAIN: Capabilities: 0xaa
I (448) pp: pp rom version: 9387209
I (448) net80211: net80211 rom version: 9387209
I (458) wifi:wifi driver task: 3fc9e6ec, prio:23, stack:6656, core=0
I (458) system_api: Base MAC address is not set
I (458) system_api: read default base MAC address from EFUSE
I (468) wifi:wifi firmware version: 9f5e450
I (468) wifi:wifi certification version: v7.0

the kernel log:

[    4.486794] esp32c3:process_esp_bootup_event: 
[    4.486794] Received ESP bootup event
[    4.486855] esp32c3:process_event_esp_bootup: EVENT: 3
[    4.486876] esp32c3:process_event_esp_bootup: EVENT: 2
[    4.486897] esp32c3:adjust_spi_clock: ESP Reconfigure SPI CLK to 30 MHz
[    4.486915] esp32c3:process_event_esp_bootup: EVENT: 0
[    4.486933] esp32c3:process_event_esp_bootup: EVENT: 4
[    4.486951] esp32c3:process_event_esp_bootup: EVENT: 1
[    4.486971] esp32c3:process_fw_data: ESP chipset's last reset cause:
[    4.486988] esp32c3:print_reset_reason: POWERON_RESET
[    4.487009] esp32c3:check_esp_version: esp32: ESP Firmware version: 1.0.2
[    4.487031] esp32c3:process_event_esp_bootup: ESP chipset detected [esp32-c3]
[    4.487103] esp32c3:add_network_iface: add_network_iface
[    4.487599] esp32c3:esp_reg_notifier: 629 esp_wifi_device not initialized yet 
[    4.487871] esp32c3:add_network_iface: add_network_iface before esp_cfg80211_add_iface
[    4.527298] esp32c3:add_network_iface: add_network_iface after esp_cfg80211_add_iface
[    4.527342] esp32c3:process_capabilities: ESP peripheral capabilities: 0xaa
[    4.527368] esp32c3:print_capabilities: Capabilities: 0xaa. Features supported are:
[    4.527386] esp32c3:print_capabilities:       * WLAN on SPI
[    4.527401] esp32c3:print_capabilities:       * BT/BLE
[    4.527418] esp32c3:print_capabilities:         - HCI over UART
[    4.527435] esp32c3:print_capabilities:         - BLE only
[    8.274566] esp32c3:cmd_auth_request: Authentication request: f4:84:8d:fb:5a:17 6 0 0 0
[    8.517273] esp32c3:cmd_assoc_request: Association request: f4:84:8d:fb:5a:17 6 22
[    8.548892] esp32c3:process_assoc_event: Connection status: 0
[    8.549930] esp32c3:process_rx_packet: Rx PACKET_TYPE_EAPOL!!!!
[    8.550514] NOHZ tick-stop error: Non-RCU local softirq work is pending, handler #08!!!
[    8.566302] esp32c3:process_rx_packet: Rx PACKET_TYPE_EAPOL!!!!
[    8.566430] NOHZ tick-stop error: Non-RCU local softirq work is pending, handler #08!!!
[   10.706334] esp32c3:esp_inetaddr_event: NETDEV_UP interface espsta0 ip changed to  192.168.001.138

So,how can I support ble over uart?

**Comment by mantriyogesh** [comment]
Hello @felix-lhp ,

## Raspberry Pi
You can check the steps in https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/setup.md#23-sdiospi--uart-configuration . These are steps for Raspberry Pi.

This documentation is for Raspberry Pi as an example. Other Linux systems also work more or less similar way, just need to port the Raspberry Pi specific things to your own Linux SoC.

## Other Linux

#### Porting guide
If you use Linux different from the Raspberry Pi, please have a look at [porting_guide](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/porting_guide.md).

#### UART hardware
- Decide you use 2-pin or 4-pin uart.
- To set up GPIOs in UART mode, Device Tree config for your platform (SoC) may need changes as per your Datasheets.

#### Base UART testing
- Also, before using Bluetooth over UART, it would be good idea to test your pure UART functionality (without Bluetooth) first.
- Test (two or four) UART pins are working fine with normal UART example (may be from ESP-IDF example on ESP chipset or any other platform user space UART application)

#### rpi_init.sh porting
You would need to set up your GPIOs in UART mode and port UART functionality in  [rpi_init.sh](https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/porting_guide.md#22-rpi_initsh)

#### Run Bluetooth over UART 
On ported UART software and hardware, you should be able to run the Bluetooth over UART similar to Raspberry Pi.

**Comment by mantriyogesh** [comment]
For UART, 'esp_init_bt()' function would not be called.
This function would be called only for:
1. `Bluetooth over SPI` _i.e._ `HCI over SPI` OR
1. `Bluetooth over SDIO` _i.e._ `HCI over SDIO`

For UART mode, the Bluetooth packets do not enter ESP-Hosted and use Linux base Bluetooth over UART kernel driver, transparently.

**Comment by mantriyogesh** [comment]
![Screenshot 2023-12-08 at 5 09 14 PM](https://github.com/espressif/esp-hosted/assets/67477199/e5eb7a90-8d70-407e-b8e9-304372136521)

**Comment by felix-lhp** [comment]
Thaks!

**Comment by mantriyogesh** [comment]
Can we close this issue?

**Comment by xmlhp** [comment]
`# insmod /oem/usr/ko/hci_uart.ko 
[ 1100.881163] Bluetooth: HCI UART driver ver 2.3
[ 1100.881191] Bluetooth: HCI UART protocol H4 registered
# hciconfig -a
# hciattach -s 921600 /dev/ttyS0 any 921600 flow nosleep
Device setup complete
# [ 1113.229270] Bluetooth: hci0: command 0x1003 tx timeout
hciconfig [ 1115.362601] Bluetooth: hci0: command 0x1001 tx timeout

hci0:   Type: Primary  Bus: UART
        BD Address: 00:00:00:00:00:00  ACL MTU: 0:0  SCO MTU: 0:0
        DOWN INIT RUNNING 
        RX bytes:0 acl:0 sco:0 events:0 errors:0
        TX bytes:12 acl:0 sco:0 commands:3 errors:0

# hciconfig [ 1117.495936] Bluetooth: hci0: command 0x1009 tx timeout
-a
hci0:   Type: Primary  Bus: UART
        BD Address: 00:00:00:00:00:00  ACL MTU: 0:0  SCO MTU: 0:0
        DOWN INIT RUNNING 
        RX bytes:0 acl:0 sco:0 events:0 errors:0
        TX bytes:12 acl:0 sco:0 commands:3 errors:0
        Features: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
        Packet type: DM1 DH1 HV1 
        Link policy: 
        Link mode: PERIPHERAL ACCEPT 

# chmod +x /oem/usr/bin/bluetoothctl 
# ifconfig -a
espsta0   Link encap:Ethernet  HWaddr C0:4E:30:84:98:B4  
          inet addr:192.168.1.129  Bcast:192.168.1.255  Mask:255.255.255.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:9699 errors:0 dropped:632 overruns:0 frame:0
          TX packets:3501 errors:0 dropped:11 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:4043671 (3.8 MiB)  TX bytes:268816 (262.5 KiB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

# hciconfig hci0 up
[ 1386.509299] Bluetooth: hci0: command 0x1003 tx timeout
[ 1388.642545] Bluetooth: hci0: command 0x1001 tx timeout
[ 1390.775887] Bluetooth: hci0: command 0x1009 tx timeout
Can't init device hci0: Connection timed out (110)
# hciconfig -a
hci0:   Type: Primary  Bus: UART
        BD Address: 00:00:00:00:00:00  ACL MTU: 0:0  SCO MTU: 0:0
        DOWN 
        RX bytes:0 acl:0 sco:0 events:0 errors:0
        TX bytes:24 acl:0 sco:0 commands:6 errors:0
        Features: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
        Packet type: DM1 DH1 HV1 
        Link policy: 
        Link mode: PERIPHERAL ACCEPT 

# `

Can't init device hci0: Connection timed out (110)

**Comment by mantriyogesh** [comment]
@xmlhp , can you please create different issue for this?

**Comment by mantriyogesh** [comment]
If you wish you can refer this issue number there.

**Comment by xmlhp** [comment]
@mantriyogesh ，this issue was create by me, felix-lhp is me.

**Comment by mantriyogesh** [comment]
Oh I get it. Sorry for the confusion.
Please provide the ESP side logs and dmesg logs at host for same test?

Is this Raspberry Pi?
If not have you configured UART on your expected GPIOs?
https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/porting_guide.md#22-rpi_initsh
rpi_init.sh has uart pin config, which need Porting as per the new Linux host.


Option 1) test basic uart on new pins to be working. Once uart is working, can reuse for bluetooth.
If uart not configured in device tree of SoC config, this will fail.
Also, to let you aware, this Bluetooth over UART is agnostic of ESP-Hosted.
Meaning, the base Linux driver transparently send bt packets through uart to bt controller.
Packets do not enter ESP-Hosted at all. So even debugging using ESP-Hosted will not be possible.
This relies on base Linux hci uart driver.

Easy option 2) if sdio or spi is already working you can use VHCI instead of uart. In this case you need to follow documentation for either 'SPI only' or 'SDIO only' transport. This way no new extra set of pins needed.
Packets will be debug able using ESP-Hosted.

**Comment by xmlhp** [comment]
Is this Raspberry Pi?
no  Raspberry Pi.linux run on arm soc.I had enable uart0(pin connect to c3) at device tree.

**Comment by xmlhp** [comment]
On our device, it is 4 pin connect  to c3 uart.I don know ,whate is the mean, bt_init()
, 15 a0 ,-->connect to RX@c3? 14 a0 -->connect to TXd@c3? pu?
bt_init()
{
    sudo raspi-gpio set 15 a0 pu
    sudo raspi-gpio set 14 a0 pu
    if [ "$BT_INIT_SET" = "4" ] ; then
        sudo raspi-gpio set 16 a3 pu
        sudo raspi-gpio set 17 a3 pu
    fi
}

That is our configurtion for uart0 (connect to c3 uart)
  /omit-if-no-ref/
        uart0m1_xfer: uart0m1-xfer {
            rockchip,pins =
                /* uart0_rx_m1 */
                <2 RK_PB0 1 &pcfg_pull_up>,
                /* uart0_tx_m1 */
                <2 RK_PB1 1 &pcfg_pull_up>;
        };

        /omit-if-no-ref/
        uart0m1_ctsn: uart0m1-ctsn {
            rockchip,pins =
                /* uart0m1_ctsn */
                <2 RK_PA7 1 &pcfg_pull_none>;
        };
        /omit-if-no-ref/
        uart0m1_rtsn: uart0m1-rtsn {
            rockchip,pins =
                /* uart0m1_rtsn */
                <2 RK_PA6 1 &pcfg_pull_none>;
        };


    uart0: serial@ff4a0000 {
        compatible = "rockchip,rv1106-uart", "snps,dw-apb-uart";
        reg = <0xff4a0000 0x100>;
        interrupts = <GIC_SPI 25 IRQ_TYPE_LEVEL_HIGH>;
        reg-shift = <2>;
        reg-io-width = <4>;
        dmas = <&dmac 7>, <&dmac 6>;
        clock-frequency = <24000000>;
        clocks = <&cru SCLK_UART0>, <&cru PCLK_UART0>;
        clock-names = "baudclk", "apb_pclk";
        pinctrl-names = "default";
        pinctrl-0 = <&uart0m1_xfer &uart0m1_ctsn &uart0m1_rtsn>;
        status = "disabled";
    };
 
&uart0{
    status = "okay";
    pinctrl-0 = <&uart0m1_xfer>;
};

**Comment by mantriyogesh** [comment]
I am not too sure about device tree.
But the clock is 24MHz, looks very high. 

pu is pull up. raspberry Pi has easy config to set pin in`a0` or`a3` alternating fiction, which will be uart with pull up.

Check the clock configs again. Is this max freq or the freq at which uart will start the clock.
How the baudrate are matched to that of clock in device tree?
Simply check the logic analyser and try some sample uart example from IDF first.

If uart works, there should be less variables in bluetooth running over UART.

**Comment by mantriyogesh** [comment]
As you had pointed correctly, bt_init() is to be ported correctly (using device tree).
Also the arg passed for two pin uart and 4 pin uart differs.

If you wish to use 2 wire uart, you have to may be disable flow control from ESP using source .
You can check IDF.py menuconfig to disable the flow control.

Unless base uart is tested fine, I wouldn't suggest to try Bluetooth on uart.

**Comment by xmlhp** [comment]
Simply check the logic analyser and try some sample uart example from IDF first.
I use the firmware you offer,because I setup compile env failed by the script@host

**Comment by mantriyogesh** [comment]
Sharing firmware and debugging is lot time consuming and confusing at times.
If you share the issues for script (not sure which), along with logs, we can look into issues and get it working.

Firmware building setup can be done on your laptop or development machine itself. 
Using host for firmware building will be overkill.

**Comment by xmlhp** [comment]
This the log @c3 statrup,the blue  uart baudrate:921600
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x1 (POWERON),boot:0xf (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd6100,len:0x16cc
load:0x403ce000,len:0x930
load:0x403d0000,len:0x2d28
entry 0x403ce000
I (30) boot: ESP-IDF v4.4.1-584-g91dc99a3ce-dirty 2nd stage bootloader
I (30) boot: compile time 17:17:27
I (30) boot: chip revision: 4
I (34) boot_comm: chip revision: 4, min. bootloader chip revision: 3
I (41) boot.esp32c3: SPI Speed      : 80MHz
I (46) boot.esp32c3: SPI Mode       : DIO
I (50) boot.esp32c3: SPI Flash Size : 4MB
I (55) boot: Enabling RNG early entropy source...
I (60) boot: Partition Table:
I (64) boot: ## Label            Usage          Type ST Offset   Length
I (71) boot:  0 nvs              WiFi data        01 02 00009000 00004000
I (79) boot:  1 otadata          OTA data         01 00 0000d000 00002000
I (86) boot:  2 phy_init         RF data          01 01 0000f000 00001000
I (94) boot:  3 factory          factory app      00 00 00010000 00100000
I (101) boot:  4 ota_0            OTA app          00 10 00110000 00100000
I (109) boot:  5 ota_1            OTA app          00 11 00210000 00100000
I (116) boot: End of partition table
I (121) boot: Defaulting to factory image
I (125) boot_comm: chip revision: 4, min. application chip revision: 3
I (132) esp_image: segment 0: paddr=00010020 vaddr=3c090020 size=17368h ( 95080) map
I (156) esp_image: segment 1: paddr=00027390 vaddr=3fc90c00 size=032f0h ( 13040) load
I (158) esp_image: segment 2: paddr=0002a688 vaddr=40380000 size=05990h ( 22928) load
I (166) esp_image: segment 3: paddr=00030020 vaddr=42000020 size=8bdd8h (572888) map
I (260) esp_image: segment 4: paddr=000bbe00 vaddr=40385990 size=0b1c8h ( 45512) load
I (268) esp_image: segment 5: paddr=000c6fd0 vaddr=50000010 size=00010h (    16) load
I (274) boot: Loaded app from partition at offset 0x10000
I (274) boot: Disabling RNG early entropy source...
I (288) cpu_start: Pro cpu up.
I (297) cpu_start: Pro cpu start user code
I (297) cpu_start: cpu freq: 160000000
I (297) cpu_start: Application information:
I (300) cpu_start: Project name:     network_adapter
I (305) cpu_start: App version:      2027593
I (310) cpu_start: Compile time:     Dec 16 2022 17:20:30
I (316) cpu_start: ELF file SHA256:  9f0bf326bb4e3f75...
I (322) cpu_start: ESP-IDF:          v4.4.1-584-g91dc99a3ce-dirty
I (329) heap_init: Initializing. RAM available for dynamic allocation:
I (336) heap_init: At 3FC980D0 len 00027F30 (159 KiB): DRAM
I (343) heap_init: At 3FCC0000 len 0001F060 (124 KiB): STACK/DRAM
I (349) heap_init: At 50000020 len 00001FE0 (7 KiB): RTCRAM
I (356) spi_flash: detected chip: generic
I (361) spi_flash: flash io: dio
I (365) sleep: Configure to isolate all GPIO pins in sleep state
I (371) sleep: Enable automatic switching of GPIO sleep configuration
I (378) coexist: coexist rom version 9387209
I (383) cpu_start: Starting scheduler.
I (388) stats: *********************************************************************
I (388) stats:                 ESP-Hosted Firmware version :: 1.0.2                        
I (398) stats:                 Transport used :: SPI + UART                    
I (408) stats: *********************************************************************
I (418) FW_MAIN: Supported features are:
I (418) FW_MAIN: - WLAN over SPI
I (428) FW_BT: - BT/BLE
I (428) FW_BT:    - HCI Over UART
I (428) FW_BT:    - BLE only
I (438) FW_MAIN: Capabilities: 0xaa
I (448) pp: pp rom version: 9387209
I (448) net80211: net80211 rom version: 9387209
I (458) wifi:wifi driver task: 3fc9e6ec, prio:23, stack:6656, core=0
I (458) system_api: Base MAC address is not set
I (458) system_api: read default base MAC address from EFUSE
I (468) wifi:wifi firmware version: 9f5e450
I (468) wifi:wifi certification version: v7.0
I (478) wifi:config NVS flash: disabled
I (478) wifi:config nano formating: disabled
I (478) wifi:Init data frame dynamic rx buffer num: 32
I (488) wifi:Init management frame dynamic rx buffer num: 32
I (488) wifi:Init management short buffer num: 32
I (498) wifi:Init dynamic tx buffer num: 32
I (498) wifi:Init static tx FG buffer num: 2
I (508) wifi:Init static rx buffer size: 1600
I (508) wifi:Init static rx buffer num: 10
I (518) wifi:Init dynamic rx buffer num: 32
I (518) wifi_init: rx ba win: 6
I (518) wifi_init: tcpip mbox: 32
I (528) wifi_init: udp mbox: 6
I (528) wifi_init: tcp mbox: 6
I (528) wifi_init: tcp tx win: 5744
I (538) wifi_init: tcp rx win: 5744
I (538) wifi_init: tcp mss: 1440
I (548) wifi_init: WiFi IRAM OP enabled
I (548) wifi_init: WiFi RX IRAM OP enabled
**I (558) gpio: GPIO[5]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (568) gpio: GPIO[19]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (568) gpio: GPIO[8]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (578) gpio: GPIO[18]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0** 
I (588) FW_BT: baud rate for HCI uart :: 921600 

I (598) BTDM_INIT: BT controller compile version [05195c9]
I (598) phy_init: phy_version 912,d001756,Jun  2 2022,16:28:07
I (638) BTDM_INIT: Bluetooth MAC: c0:4e:30:84:98:b6

I (648) FW_MAIN: ESP Bluetooth MAC addr: c0-4e-30-84-98-b6

I (648) FW_SPI: Using SPI interface
I (648) gpio: GPIO[3]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (658) gpio: GPIO[4]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1668) FW_MAIN: Initial set up done
I (1678) FW_MAIN: INIT Interface command

I (1678) wifi:mode : sta (c0:4e:30:84:98:b4)
I (1678) wifi:enable tsf
I (1678) FW_CMD: Wifi Sta mode set

I (1698) FW_MAIN: Get MAC command
# hciattach -s 921600 /dev/ttyS0 any 921600 flow nosleep
[  252.644114] of_dma_request_slave_channel: dma-names property of node '/serial@ff4a0000' missing or empty
Device setup complete
[  252.644145] dw-apb-uart ff4a0000.serial: failed to request DMA, use interrupt mode
# [  254.776209] Bluetooth: hci0: command 0x1003 tx timeout
# hciattach -s 921600 /dev/ttyS0 any 921600 flow nosleep
Can't set device: Device or resource busy
Can't initialize device: Device or resource busy
# [  256.909555] Bluetooth: hci0: command 0x1001 tx timeout
[  259.042898] Bluetooth: hci0: command 0x1009 tx timeout


From the log. gpio5,gpio19 .gpio8,gpio18 ,what is function of theses gpios?

This is the operat log @arm device
# hciconfig -a
hci0:   Type: Primary  Bus: UART
        BD Address: 00:00:00:00:00:00  ACL MTU: 0:0  SCO MTU: 0:0
        DOWN 
        RX bytes:0 acl:0 sco:0 events:0 errors:0
        TX bytes:12 acl:0 sco:0 commands:3 errors:0
        Features: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
        Packet type: DM1 DH1 HV1 
        Link policy: 
        Link mode: PERIPHERAL ACCEPT 

# hciconfig hci0 up
[  283.789546] Bluetooth: hci0: command 0x1003 tx timeout
[  285.922864] Bluetooth: hci0: command 0x1001 tx timeout


[  288.056205] Bluetooth: hci0: command 0x1009 tx timeout
^C
# hciconfig -a
hci0:   Type: Primary  Bus: UART
        BD Address: 00:00:00:00:00:00  ACL MTU: 0:0  SCO MTU: 0:0
        DOWN 
        RX bytes:0 acl:0 sco:0 events:0 errors:0
        TX bytes:24 acl:0 sco:0 commands:6 errors:0
        Features: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
        Packet type: DM1 DH1 HV1 
        Link policy: 
        Link mode: PERIPHERAL ACCEPT 

#

**Comment by mantriyogesh** [comment]
Are you able to.verify the base UART example with your Linux UART?
try some user space uart verification example on Linux along with base UART example from ESP-IDF? https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-reference/peripherals/uart.html

I suspect your Linux is not correctly configured UART.
https://github.com/espressif/esp-hosted/issues/295#issuecomment-1888684148

**Comment by mantriyogesh** [comment]
Have you changed anything at slave?

**Comment by mantriyogesh** [comment]
> what is function of theses gpios

Pin connections are documented here: 
https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/Linux_based_host/UART_setup.md

**Comment by xmlhp** [comment]
Can TX, RX, CTS, RTS on C3 be reconfigured pins?

**Comment by mantriyogesh** [comment]
> Can TX, RX, CTS, RTS on C3 be reconfigured pins?
Yes, they are reconfigurable in Slave at
https://github.com/espressif/esp-hosted/blob/f32dc7495eb79891d667c8ce044f36cbdff22a70/esp_hosted_ng/esp/esp_driver/network_adapter/main/include/slave_bt.h#L82-L85
Just re-flash the c3. You can also add a log in NG to see new pins used using:
https://github.com/espressif/esp-hosted/blob/f32dc7495eb79891d667c8ce044f36cbdff22a70/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c#L296-L297

I assume, you are still at old commit or using release binary.
Please use latest master at both ESP and host sides.
If the git commit used is different at ESP and host, you may fall into the issues.

**Comment by xmlhp** [comment]
I'm going to swap the configuration of CTS, RTS.Is that OK?
#define BT_RTS_PIN		8
 #define BT_CTS_PIN		19

**Comment by mantriyogesh** [comment]
Also, if you check,
https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/Linux_based_host/UART_setup.md#12-raspberry-pi-software-setup

## Set the GPIO Pins to UART mode
Here, in Raspberry Pi, simple config to enable uart is:
enable_uart=1

Similarly, you need to configure the device tree (if any easy config available, try that) so that the UART hardware peripheral is ready 

## If present, disable native bluetooth to avoid confusion
As raspberry Pi own has bluetooth, we disable the native bluetooth using:
dtoverlay=disable-bt

This will make sure the native bluetooth will not start. ESP-Hosted provided bluetooth anyway will work over HCI standard and will be detected fine.
If we do not disable bluetooth as above, you will see two bluetooth agents.
If your SoC doesn't have the native bluetooth, ignore this step.


## Make sure the uart bus is unused for any other purpose.
As we use uart pins, where system console by default is sent, we make sure the system console is not sent to the uart instance we wish to use for bluetooth.

For that purpose, We remove 'serial0' from the /boot/cmdline.txt in case of Raspberry Pi.
You ensure that the uart device enabled, is not getting used by anyone else in kernel or user space.
Also mark which device is exposed by this.

## Use correct uart device name
Easy way to check, if /dev/XYZ is created after enable of uart in device tree and destroyed when you disable the UART from device tree, /dev/XYZ is the device you should be interested in.

## Test all UART lines
Do not skip to bluetooth yet. try the simple ESP-IDF UART example first with you Linux uaser space simple uart application.
Refer https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-reference/peripherals/uart.html for simple UART examples in ESP-IDF

If your UART lines are correct then we can try Bluetooth over UART.

## Use exact baud rate, at which ESP is configured. If baud rate mismatched, nothing will work.
You can check the clock etc using logic analyzer.

**Comment by mantriyogesh** [comment]
> https://github.com/espressif/esp-hosted/issues/295#issuecomment-1891233094

Yes is okay.

**Comment by mantriyogesh** [comment]
Also, following is Raspberry Pi specific link, but verify if you have all required software and permission available:

https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/Linux_based_host/Troubleshoot.md#31-bluetooth-does-not-work

**Comment by xmlhp** [comment]
Host-ng ,To compile the firmware for c3, refer to this page?
https://github.com/espressif/esp-hosted/tree/master/esp_hosted_ng/esp/esp_driver

**Comment by mantriyogesh** [comment]
Yes. thats correct.

The pins could be changed at : https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/esp/esp_driver/network_adapter/main/include/slave_bt.h#L82

**Comment by mantriyogesh** [comment]
Please post output of:
```
$ uname -a
```

**Comment by xmlhp** [comment]
@device :5.10

@ linux Development Environment
Linux ubuntu-server-22-04 5.15.0-91-generic #101-Ubuntu SMP Tue Nov 14 13:30:08 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

**Comment by mantriyogesh** [comment]
I also tried 2 line uart solution as well, but using FG. Similar changes can be done to NG also and try for 2 wire (only Tx and RX)

```diff
 I  ~/c/h6   *$…  esp_hosted_fg/esp/esp_driver/network_adapter  git diff main/slave_bt.c                                                                                                                                                  443ms  Mon Jan 15 12:23:02 2024
diff --git a/esp_hosted_fg/esp/esp_driver/network_adapter/main/slave_bt.c b/esp_hosted_fg/esp/esp_driver/network_adapter/main/slave_bt.c
index 4699e8c3..b374ccaf 100644
--- a/esp_hosted_fg/esp/esp_driver/network_adapter/main/slave_bt.c
+++ b/esp_hosted_fg/esp/esp_driver/network_adapter/main/slave_bt.c
@@ -306,8 +306,8 @@ static void init_uart_c3_s3(void)
         .data_bits = UART_DATA_8_BITS,
         .parity = UART_PARITY_DISABLE,
         .stop_bits = UART_STOP_BITS_1,
-        .flow_ctrl = UART_HW_FLOWCTRL_CTS_RTS,
-        .rx_flow_ctrl_thresh = UART_RX_THRS,
+        .flow_ctrl = UART_HW_FLOWCTRL_DISABLE,
+        //.rx_flow_ctrl_thresh = UART_RX_THRS,
         .source_clk = UART_SCLK_APB,
     };
     ESP_ERROR_CHECK(uart_param_config(BLUETOOTH_UART, &uart_config));
@@ -370,9 +370,8 @@ static void init_uart_esp32(void)


        ESP_ERROR_CHECK( uart_set_pin(BLUETOOTH_UART, BT_TX_PIN,
-               BT_RX_PIN, BT_RTS_PIN, BT_CTS_PIN) );
-       ESP_LOGI(BT_TAG, "UART Pins: Tx:%u Rx:%u RTS:%u CTS:%u",
-                       BT_TX_PIN, BT_RX_PIN, BT_RTS_PIN, BT_CTS_PIN);
+               BT_RX_PIN, UART_PIN_NO_CHANGE, UART_PIN_NO_CHANGE) );
+       ESP_LOGI(BT_TAG, "UART Pins: Tx:%u Rx:%u", BT_TX_PIN, BT_RX_PIN);

 }
 #endif
```

Two pin uart solution, rpi_init.sh command should use `btuart_2pins`
```sh
$ bash rpi_init.sh spi btuart_2pins
```

With 
```sh
$ sudo hciattach -s 921600 /dev/serial0 any 921600 flow
```
should get 

```sh 
kysoh@raspberrypi:~/h6/esp_hosted_fg/host/linux/host_control $ hciconfig
hci0:	Type: Primary  Bus: UART
	BD Address: 70:04:1D:14:FD:06  ACL MTU: 251:12  SCO MTU: 0:0
	UP RUNNING
	RX bytes:874 acl:0 sco:0 events:73 errors:0
	TX bytes:603 acl:0 sco:0 commands:73 errors:0
```


But anyway, base assumption is very clear, the UART should work first reliably, then can use Bluetooth on UART.

**Comment by xmlhp** [comment]
@ https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/setup.md#2-transport-layer-configuration
`For ESP32-C3, select chip revision in addition. Navigate to Component config → ESP32C3-Specific → Minimum Supported ESP32-C3 Revision and select chip version of ESP32-C3.`
But I don`t find **ESP32C3-Specific** in **ESP32C3-Specific**?

**Comment by xmlhp** [comment]
Sorry, It's wrong,It should be:
But I don`t find **ESP32C3-Specific** in  **Component config**?

**Comment by xmlhp** [comment]
I can not use idy.py burning, because the compiler of the computer has no serial port.
I want to copy the firmware to the window notebook and burn it with SPIDownload.

Is THE FOLLOWING a FEW FILES copy to the window NOTEBOOK, with SPIDowload(@flash_dowload_tool) BURN into esp32c3?
esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/build/bootloader/bootloader.bin
esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/build/partition_table/partition-table.bin
esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/build/ota_data_initial.bin
esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/build/network_adapter.bin

**Comment by xmlhp** [comment]
This is because our development environment is on a public server. Files can only be downloaded via the Internet.

**Comment by xmlhp** [comment]
This is the log of that firmware build by me:
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x1 (POWERON),boot:0x7 (DOWNLOAD(USB/UART0/1))
waiting for download

ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x1 (POWERON),boot:0xf (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd5820,len:0x170c
load:0x403cc710,len:0x968
load:0x403ce710,len:0x2f68
entry 0x403cc710
I (30) boot: ESP-IDF d3c99ed3-dirty 2nd stage bootloader
I (30) boot: compile time Jan 15 2024 05:41:10
I (30) boot: chip revision: v0.4
I (34) boot.esp32c3: SPI Speed      : 80MHz
I (39) boot.esp32c3: SPI Mode       : DIO
I (43) boot.esp32c3: SPI Flash Size : 4MB
I (48) boot: Enabling RNG early entropy source...
I (54) boot: Partition Table:
I (57) boot: ## Label            Usage          Type ST Offset   Length
I (64) boot:  0 nvs              WiFi data        01 02 00009000 00004000
I (72) boot:  1 otadata          OTA data         01 00 0000d000 00002000
I (79) boot:  2 phy_init         RF data          01 01 0000f000 00001000
I (87) boot:  3 factory          factory app      00 00 00010000 00100000
I (94) boot:  4 ota_0            OTA app          00 10 00110000 00100000
I (102) boot:  5 ota_1            OTA app          00 11 00210000 00100000
I (109) boot: End of partition table
I (114) boot: Defaulting to factory image
I (118) esp_image: segment 0: paddr=00010020 vaddr=3c090020 size=1ff48h (130888) map
I (148) esp_image: segment 1: paddr=0002ff70 vaddr=3fc93e00 size=000a8h (   168) load
I (148) esp_image: segment 2: paddr=00030020 vaddr=42000020 size=871cch (553420) map
I (242) esp_image: segment 3: paddr=000b71f4 vaddr=3fc93ea8 size=03004h ( 12292) load
I (244) esp_image: segment 4: paddr=000ba200 vaddr=40380000 size=13c0ch ( 80908) load
I (269) boot: Loaded app from partition at offset 0x10000
I (269) boot: Disabling RNG early entropy source...
I (280) cpu_start: Unicore app
I (280) cpu_start: Pro cpu up.
I (289) cpu_start: Pro cpu start user code
I (289) cpu_start: cpu freq: 160000000 Hz
I (289) cpu_start: Application information:
I (292) cpu_start: Project name:     network_adapter
I (298) cpu_start: App version:      release/ng-v1.0.2-155-gf32dc74-
I (305) cpu_start: Compile time:     Jan 15 2024 05:41:08
I (311) cpu_start: ELF file SHA256:  8c103cc1b980825a...
I (317) cpu_start: ESP-IDF:          d3c99ed3-dirty
I (323) cpu_start: Min chip rev:     v0.3
I (327) cpu_start: Max chip rev:     v0.99 
I (332) cpu_start: Chip rev:         v0.4
I (337) heap_init: Initializing. RAM available for dynamic allocation:
I (344) heap_init: At 3FC9A870 len 00041EA0 (263 KiB): DRAM
I (350) heap_init: At 3FCDC710 len 00002950 (10 KiB): STACK/DRAM
I (357) heap_init: At 50000010 len 00001FD8 (7 KiB): RTCRAM
I (364) spi_flash: detected chip: generic
I (368) spi_flash: flash io: dio
I (372) sleep: Configure to isolate all GPIO pins in sleep state
I (379) sleep: Enable automatic switching of GPIO sleep configuration
I (386) coexist: coex firmware version: 80b0d89
I (391) coexist: coexist rom version 9387209
I (396) app_start: Starting scheduler on CPU0
I (401) main_task: Started on CPU0
I (401) main_task: Calling app_main()
I (401) stats: *********************************************************************
I (411) stats:                 ESP-Hosted Firmware version :: 1.0.3                        
I (421) stats:                 Transport used :: SPI only                      
I (431) stats: *********************************************************************
I (441) FW_MAIN: Supported features are:
I (441) FW_MAIN: - WLAN over SPI
I (451) FW_BT: - BT/BLE
I (451) FW_BT:    - **HCI Over SPI**
I (451) FW_BT:    - BLE only
I (461) FW_MAIN: Capabilities: 0xe8
pp rom version: 9387209
net80211 rom version: 9387209
I (481) wifi:wifi driver task: 3fca1000, prio:23, stack:6656, core=0
I (481) wifi:wifi firmware version: fddc5e5
I (481) wifi:wifi certification version: v7.0
I (481) wifi:config NVS flash: disabled
I (481) wifi:config nano formating: disabled
I (491) wifi:Init data frame dynamic rx buffer num: 32
I (491) wifi:Init management frame dynamic rx buffer num: 32
I (501) wifi:Init management short buffer num: 32
I (501) wifi:Init dynamic tx buffer num: 32
I (511) wifi:Init static tx FG buffer num: 2
I (511) wifi:Init static rx buffer size: 1600
I (511) wifi:Init static rx buffer num: 10
I (521) wifi:Init dynamic rx buffer num: 32
I (521) wifi_init: rx ba win: 6
I (531) wifi_init: tcpip mbox: 32
I (531) wifi_init: udp mbox: 6
I (531) wifi_init: tcp mbox: 6
I (541) wifi_init: tcp tx win: 5744
I (541) wifi_init: tcp rx win: 5744
I (551) wifi_init: tcp mss: 1440
I (551) wifi_init: WiFi IRAM OP enabled
I (551) wifi_init: WiFi RX IRAM OP enabled
I (561) BLE_INIT: BT controller compile version [963cad4]
I (561) phy_init: phy_version 970,1856f88,May 10 2023,17:44:12
W (571) phy_init: failed to load RF calibration data (0xffffffff), falling back to full calibration
I (641) BLE_INIT: Bluetooth MAC: c0:4e:30:84:98:b6

I (651) FW_MAIN: ESP Bluetooth MAC addr: c0-4e-30-84-98-b6

I (651) FW_SPI: Using SPI interface
I (651) gpio: GPIO[3]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (661) gpio: GPIO[4]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1671) FW_MAIN: Initial set up done
I (1671) main_task: Returned from app_main()
I (1691) FW_MAIN: INIT Interface command

Blue ove spi？How do confiuriat, after idf.py menuconfig?

**Comment by xmlhp** [comment]
How to configure @ Componentconfig->Bluetooth->Controller Options?

**Comment by mantriyogesh** [comment]
Bluetooth over SPI should be relatively simpler, if your SPI is correctly setup already.
Refer to https://github.com/espressif/esp-hosted/tree/master/esp_hosted_ng ->   [2. Hardware and Software setup](https://github.com/espressif/esp-hosted/tree/master/esp_hosted_ng#2-hardware-and-software-setup)

**Comment by xmlhp** [comment]
We using uart for bluethooh.

**Comment by xmlhp** [comment]
We are to use the serial port to control Bluetooth. How do you configure this

**Comment by mantriyogesh** [comment]
Your firmware built is 
```
I (451) FW_BT: - HCI Over SPI
```
If you want HCI over UART, please follow UART documentation and make changes accordingly using `idf.py menuconfig`

**Comment by mantriyogesh** [comment]
Please read and follow, https://github.com/espressif/esp-hosted/issues/295#issuecomment-1891431765

**Comment by xmlhp** [comment]
Afer I select UART for bluetooth by  **Component config -> Bluetooth -> Bluetooth controller -> HCI UART(H4) Options ->     UART Baudrate for HCI**
idf.py build error:

I```
n file included from /home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/main/app_main.c:47:
/home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c: In function 'init_uart':
/home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c:277:30: error: 'GPIO_PIN_INTR_DISABLE' undeclared (first use in this function); did you mean 'GPIO_INTR_DISABLE'?
  277 |                 .intr_type = GPIO_PIN_INTR_DISABLE,     //disable interrupt
      |                              ^~~~~~~~~~~~~~~~~~~~~
      |                              GPIO_INTR_DISABLE
/home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c:277:30: note: each undeclared identifier is reported only once for each function it appears in
In file included from /home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c:27:
/home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c:327:54: error: 'GDMA_TRIG_PERIPH_UART' undeclared (first use in this function); did you mean 'GDMA_TRIG_PERIPH_RMT'?
  327 |         gdma_connect(s_tx_channel, GDMA_MAKE_TRIGGER(GDMA_TRIG_PERIPH_UART, 0));
      |                                                      ^~~~~~~~~~~~~~~~~~~~~
/home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/esp-idf/components/esp_hw_support/include/esp_private/gdma.h:111:34: note: in definition of macro 'GDMA_MAKE_TRIGGER'
  111 |     (gdma_trigger_t) { .periph = peri, .instance_id = SOC_##peri##id }
      |                                  ^~~~
/home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/esp-idf/components/esp_hw_support/include/esp_private/gdma.h:111:55: error: 'SOC_GDMA_TRIG_PERIPH_UART0' undeclared (first use in this function); did you mean 'SOC_GDMA_TRIG_PERIPH_UHCI0'?
  111 |     (gdma_trigger_t) { .periph = peri, .instance_id = SOC_##peri##id }
      |                                                       ^~~~
/home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c:327:36: note: in expansion of macro 'GDMA_MAKE_TRIGGER'
  327 |         gdma_connect(s_tx_channel, GDMA_MAKE_TRIGGER(GDMA_TRIG_PERIPH_UART, 0));
      |                                    ^~~~~~~~~~~~~~~~~
[887/902] Building C object esp-idf/main/CMakeFiles/__idf_main.dir/cmd.c.objninja: build stopped: subcommand failed.
ninja failed with exit code 1, output of the command is in the /home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/build/log/idf_py_stderr_output_23294 and /home/hongli/esp32/esp-hosted/esp_hosted_ng/esp/esp_driver/network_adapter/build/log/idf_py_stdout_output_23294

```

**Comment by mantriyogesh** [comment]
@kapilkedawat ++

**Comment by xmlhp** [comment]
[sdkconfig.zip](https://github.com/espressif/esp-hosted/files/13936038/sdkconfig.zip)
This is sdkconfig file @network_adapter

**Comment by xmlhp** [comment]
@mantriyogesh @kapilkedawat ,
How to fix it

**Comment by mantriyogesh** [comment]
```diff
diff --git a/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c b/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c
index c8cac4e3..3eb8cae8 100644
--- a/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c
+++ b/esp_hosted_ng/esp/esp_driver/network_adapter/main/slave_bt.c
@@ -274,7 +274,7 @@ static void init_uart(void)
        periph_module_reset(PERIPH_UHCI0_MODULE);

        gpio_config_t io_output_conf = {
-               .intr_type = GPIO_PIN_INTR_DISABLE,     //disable interrupt
+               .intr_type = DISABLE_INTR_ON_GPIO,      //disable interrupt
                .mode = GPIO_MODE_OUTPUT,       // output mode
                .pin_bit_mask = GPIO_OUTPUT_PIN_SEL,    // bit mask of the output pins
                .pull_down_en = 0,      // disable pull-down mode
@@ -324,8 +324,8 @@ static void init_uart(void)
        };
        ESP_ERROR_CHECK(gdma_new_channel(&rx_channel_config, &s_rx_channel));

-       gdma_connect(s_tx_channel, GDMA_MAKE_TRIGGER(GDMA_TRIG_PERIPH_UART, 0));
-       gdma_connect(s_rx_channel, GDMA_MAKE_TRIGGER(GDMA_TRIG_PERIPH_UART, 0));
+       gdma_connect(s_tx_channel, GDMA_MAKE_TRIGGER(GDMA_TRIG_PERIPH_UHCI, 0));
+       gdma_connect(s_rx_channel, GDMA_MAKE_TRIGGER(GDMA_TRIG_PERIPH_UHCI, 0));

        gdma_strategy_config_t strategy_config = {
                .auto_update_desc = false,
```

Does this solve the build issue?

**Comment by mantriyogesh** [comment]
@xmlhp 
Can you please share the idf version you are using?

**Comment by xmlhp** [comment]
I setup esp-idf refer this gudie:
run cmake . to setup enviornment, it will setup esp-idf as submodule to be used by network_adapter

How to find the version of idf?

**Comment by xmlhp** [comment]
Now,I can get hci0.But exec hciconfig -a, print error **Can't read local name on hci0: Input/output error (5)**
# hciattach -s 921600 /dev/ttyS0 any 921600 flow
[   36.909563] dw-apb-uart ff4a0000.serial: got rx and tx dma channels
Device setup complete
# ps -ef|grep hicattach
ps: invalid option -- e
BusyBox v1.27.2 (2023-12-14 07:24:49 UTC) multi-call binary.

Usage: ps 
# ps |grep hicattach
  312 root      1204 S    grep hicattach
# hciconfig -a
hci0:   Type: Primary  Bus: UART
        BD Address: C0:4E:30:84:98:B6  ACL MTU: 251:12  SCO MTU: 0:0
        DOWN 
        RX bytes:260 acl:0 sco:0 events:20 errors:0
        TX bytes:113 acl:0 sco:0 commands:20 errors:0
        Features: 0x00 0x00 0x00 0x00 0x60 0x00 0x00 0x00
        Packet type: DM1 DH1 HV1 
        Link policy: 
        Link mode: PERIPHERAL ACCEPT 

# hciconfig hci0 up
# hcicofig -a
-/bin/sh: hcicofig: not found
# hciconfig -a
hci0:   Type: Primary  Bus: UART
        BD Address: C0:4E:30:84:98:B6  ACL MTU: 251:12  SCO MTU: 0:0
        UP RUNNING 
        RX bytes:520 acl:0 sco:0 events:40 errors:0
        TX bytes:226 acl:0 sco:0 commands:40 errors:0
        Features: 0x00 0x00 0x00 0x00 0x60 0x00 0x00 0x00
        Packet type: DM1 DH1 HV1 
        Link policy: 
        Link mode: PERIPHERAL ACCEPT 
Can't read local name on hci0: Input/output error (5)
# hciconfig -a
hci0:   Type: Primary  Bus: UART
        BD Address: C0:4E:30:84:98:B6  ACL MTU: 251:12  SCO MTU: 0:0
        UP RUNNING 
        RX bytes:527 acl:0 sco:0 events:41 errors:0
        TX bytes:230 acl:0 sco:0 commands:41 errors:0
        Features: 0x00 0x00 0x00 0x00 0x60 0x00 0x00 0x00
        Packet type: DM1 DH1 HV1 
        Link policy: 
        Link mode: PERIPHERAL ACCEPT 
Can't read local name on hci0: Input/output error (5)
#

**Comment by xmlhp** [comment]
I refer to this page, to setup GATT server
https://github.com/espressif/esp-hosted/blob/master/esp_hosted_ng/docs/bluetooth.md
# bluetoothctl 
Waiting to connect to bluetoothd...dbus[378]: arguments to dbus_connection_get_object_path_data() were incorrect, assertion "connection != NULL" failed in file dbus-connection.c line 5905.
This is normally a bug in some application using the D-Bus library.

  D-Bus not compiled with backtrace support so unable to print a backtrace
Aborted (core dumped)
# hciconfig 
hci0:   Type: Primary  Bus: UART
        BD Address: C0:4E:30:84:98:B6  ACL MTU: 251:12  SCO MTU: 0:0
        UP RUNNING 
        RX bytes:548 acl:0 sco:0 events:44 errors:0
        TX bytes:242 acl:0 sco:0 commands:44 errors:0

# hciconfig -a
hci0:   Type: Primary  Bus: UART
        BD Address: C0:4E:30:84:98:B6  ACL MTU: 251:12  SCO MTU: 0:0
        UP RUNNING 
        RX bytes:548 acl:0 sco:0 events:44 errors:0
        TX bytes:242 acl:0 sco:0 commands:44 errors:0
        Features: 0x00 0x00 0x00 0x00 0x60 0x00 0x00 0x00
        Packet type: DM1 DH1 HV1 
        Link policy: 
        Link mode: PERIPHERAL ACCEPT 
Can't read local name on hci0: Input/output error (5)

**Comment by shreyassheth1** [comment]
https://github.com/espressif/esp-hosted/issues/295#issuecomment-1893057038
Were you able to resolve the firmware building issue by making these changes?
Also, please share logs when new firmware is flashed if you are still facing the error.

**Comment by mantriyogesh** [comment]
Check the bluetooth service running. Do you have BlueZ installed and configured correctly?

BlueZ, pulseaudio etc installation, config is out of scope for ESP-Hosted. you need to check..
You have got the base bluetooth over UART..

**Comment by xmlhp** [comment]
I had exec dbus-daemon, Now blocking on bluetoothctl
  
555 root      1912 S    dbus-daemon --config-file=/usr/share/dbus-1/system.c
  556 root      1208 R    ps
# bluetoothctl s
services/         session.conf      system-services/  system.conf
# bluetoothctl 

# hciconfig -a
hci0:   Type: Primary  Bus: UART
        BD Address: C0:4E:30:84:98:B6  ACL MTU: 251:12  SCO MTU: 0:0
        UP RUNNING 
        RX bytes:555 acl:0 sco:0 events:45 errors:0
        TX bytes:282 acl:0 sco:0 commands:54 errors:0
        Features: 0x00 0x00 0x00 0x00 0x60 0x00 0x00 0x00
        Packet type: DM1 DH1 HV1 
        Link policy: 
        Link mode: PERIPHERAL ACCEPT 
Can't read local name on hci0: Connection timed out (110)
# hciconfig -a[13597.707012] Bluetooth: hci0: command 0x0c14 tx timeout
# bluetoothctl

**Comment by xmlhp** [comment]
export                                            Print environment variables
[bluetooth]# bluetoothd[598]: src/advertising.c:register_advertisement() RegisterAdvertisement
bluetoothd[598]: src/advertising.c:client_create() Adding proxy for /org/bluez/advertising
[bluetooth]# bluetoothd[598]: src/advertising.c:register_advertisement() Registered advertisement at path /org/bluez/advertising
bluetoothd[598]: src/advertising.c:parse_discoverable() Adding Flags 0x02
bluetoothd[598]: src/advertising.c:refresh_legacy_adv() Refreshing advertisement: /org/bluez/advertising
bluetoothd[598]: src/shared/mgmt.c:send_request() [0x0000] command 0x003e
bluetoothd[598]: src/shared/mgmt.c:can_read_data() [0x0000] command 0x003e complete: 0x00
bluetoothd[598]: src/advertising.c:add_adv_callback() Advertisement registered: /org/bluez/advertising
[CHG] Controller C0:4E:30:84:98:B6 SupportedInstances: 0x05 (5)
[CHG] Controller C0:4E:30:84:98:B6 ActiveInstances: 0x01 (1)
Advertising object registered
**Tx Power: off
LocalName: textxxxx**
Appearance: off
Discoverable: on

But I'm on my phone and I'm not searching for an AD pack

**Comment by mantriyogesh** [comment]
Can you please explain the issue?

Are you able to run :
```
$ sudo bluetoothctrl
```

As I said earlier, the Bluetooth over UART is transparent mode.
Meaning, the packets do not enter ESP-Hosted.

So if any issues, you might need to check your Bluetoothd/BlueZ config, btmon when issue happens and relate with your SoC's hci_uart driver.

**Comment by xmlhp** [comment]
# ps |grep blue
  598 root      5212 S    bluetoothd -d -n
  618 root      1204 S    grep blue
# ps |grep dbus
  555 root      1920 S    dbus-daemon --config-file=/usr/share/dbus-1/system.c
  620 root      1204 S    grep dbus
# cat /proc/555/cmdline 
# **bluetoothctl** 
bluetoothd[598]: src/agent.c:add_default_agent() Default agent set to :1.5 /org/bluez/agent
bluetoothd[598]: src/adapter.c:set_mode() sending set mode command for index 0
bluetoothd[598]: src/agent.c:agent_ref() 0x623bd0: ref=1
bluetoothd[598]: src/agent.c:register_agent() agent :1.5
bluetoothd[598]: src/shared/mgmt.c:send_request() [0x0000] command 0x0009
bluetoothd[598]: src/shared/mgmt.c:can_read_data() [0x0000] command 0x0009 complete: 0x00
bluetoothd[598]: src/adapter.c:new_settings_callback() Settings: 0x00000a10
bluetoothd[598]: src/adapter.c:settings_changed() Changed settings: 0x00000010
bluetoothd[598]: src/adapter.c:settings_changed() Pending settings: 0x00000000
bluetoothd[598]: src/shared/mgmt.c:send_request() [0x0000] command 0x0018
bluetoothd[598]: src/shared/mgmt.c:can_read_data() [0x0000] command 0x0018 complete: 0x00
Agent registered
[CHG] Controller C0:4E:30:84:98:B6 Pairable: yes

I can run bluetoothctl

**Comment by mantriyogesh** [comment]
Log at ESP?
tried  https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/docs/Linux_based_host/Troubleshoot.md#32-in-case-of-bluetooth-over-uart-getting-timeouts-for-tx ?

kill and restart hciattach  helps?

**Comment by mantriyogesh** [comment]
Can we close this issue @felix-lhp ?

**Comment by xmlhp** [comment]
OK ,close this issue
